src/lib/scheduler/node_queue_scheduler.cpp

```diff
@@ -73,10 +79,22 @@ void NodeQueueScheduler::wait_for_all_tasks() {
       break;
     }
 
+    Assert(progressless_loop_count < 1'000, "Timeout: no progress while waiting for all scheduled tasks to be processed.");
+
+    if (previous_finished_task_count < num_finished_tasks) {
+      progressless_loop_count = 0;
+    }
+
     std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    ++progress_loop_count;
+    previousless_finished_task_count = num_finished_tasks;
```

Comment:
isn't this equivalent to the code below or am I missing sth?
```c++
auto progressless_loop_count = size_t{0};
auto previous_finished_task_count = size_t{0};

while (true) {
  ...
  if (previous_finished_task_count == num_finished_tasks) {
    ++progressless_loop_count;
  } else {
    progressless_loop_count = 0;
  }
  previousless_finished_task_count = num_finished_tasks;
  ...
}
```
also `num_` vs `_count`